src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,712 @@
+
+class LibraryFileSyncStatusSynchronizedTest : public LibraryFileSyncTest {
+  protected:
+    explicit LibraryFileSyncStatusSynchronizedTest(const QFileInfo& testFile)
+            : LibraryFileSyncTest(testFile) {
+    }
+
+    TrackPointer prepareTestTrack() const override {
+        TrackPointer pTrack;
+        loadTrackFromDatabaseAndVerifySourceSyncStatus(
+                &pTrack,
+                mixxx::TrackRecord::SourceSyncStatus::Synchronized);
+        return pTrack;
+    }
+
+    void checkTrackRecordSourceSyncStatus() const {
+        LibraryFileSyncTest::checkTrackRecordSourceSyncStatus(
+                mixxx::TrackRecord::SourceSyncStatus::Synchronized);
+    }
+};
+
+class LibraryFileWithMetadataSyncStatusSynchronizedTest
+        : public LibraryFileSyncStatusSynchronizedTest {
+  public:
+    LibraryFileWithMetadataSyncStatusSynchronizedTest()
+            : LibraryFileSyncStatusSynchronizedTest(kTestFileWithMetadata) {
+    }
+};
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, checkTrackRecordSourceSyncStatus) {
+    LibraryFileSyncStatusSynchronizedTest::checkTrackRecordSourceSyncStatus();
+}
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncEnabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithMetadata, true);
+}
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncDisabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithMetadata, false);
+}
+
+class LibraryFileWithoutMetadataSyncStatusSynchronizedTest
+        : public LibraryFileSyncStatusSynchronizedTest {
+  public:
+    LibraryFileWithoutMetadataSyncStatusSynchronizedTest()
+            : LibraryFileSyncStatusSynchronizedTest(kTestFileWithoutMetadata) {
+    }
+};
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, checkTrackRecordSourceSyncStatus) {
+    LibraryFileSyncStatusSynchronizedTest::checkTrackRecordSourceSyncStatus();
+}
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncEnabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithoutMetadata, true);
+}
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncDisabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithoutMetadata, false);
+}
+
+class LibraryFileSyncStatusOutdatedTest : public LibraryFileSyncTest {
+  public:
+    explicit LibraryFileSyncStatusOutdatedTest(const QFileInfo& testFile)
+            : LibraryFileSyncTest(testFile) {
+    }
+
+  protected:
+    TrackPointer prepareTestTrack() const override {
+        TrackPointer pTrack;
+        establishSourceSyncStatus(&pTrack);
+        return pTrack;
+    }
+
+    void checkTrackRecordSourceSyncStatus() const {
+        LibraryFileSyncTest::checkTrackRecordSourceSyncStatus(
+                mixxx::TrackRecord::SourceSyncStatus::Outdated);
+    }
+
+  private:
+    void establishSourceSyncStatus(TrackPointer* pLoadedTrack) const {
```

Comment:
No. This is already part of the class name. The member functions are named identical in all derived classes.

Reply:
That's no reason for giving private member function bad names.

Reply:
I disagree and will not change the naming of the member functions. This would affect all other member functions too, would produce a lot of noise and would also change the consistent implementation code of the individual tests. Currently these tests can be created simply by copy&paste because of the identical naming!

Reply:
copy & past is evil :-)  

Seriously, that is no reason for bad function names!
You may consider to manual inline them as an alternative. They are called from one place only anyway.

Reply:
Daniel, seriously. Then please rewrite everything in your style and don't bother me anymore.

I am well aware that the code is still far from ideal. I have already spent a lot of time to come up with these generic testing functions. On the other hand you are asking me to remove the generic functions and replace everything with even more duplicated, boilerplate code [here](https://github.com/mixxxdj/mixxx/pull/4628#issuecomment-1016829290).

Reply:
@uklotzde has already invested a lot of work into fixing these complex problems that have real consequences for users. Please do not bother him about implementation details that have literally no impact on people running the code.

Reply:
This functions does not longer exist. It has been replaced by the polymorphic `TrackPointer prepareTestTrack()`.

https://github.com/mixxxdj/mixxx/pull/4628/commits/a4f550ba80c3e2af1058bf5f4bcedf8cd5df1186